### PR TITLE
feat: [FC-7879] implement get_user_dates() method to get all user dates for a course#326

### PR DESCRIPTION
**Pull request by kyrylo-kh**
> NOTE: Depends on #343

## Description
This PR adds a new API function `get_user_dates(course_id, user_id)` to return a comprehensive map of dates for a given course and user.

### Functionality
* Returns a dictionary where keys are `(block_key, field)` tuples and values are `datetime` objects.
* Prioritizes user-specific overrides (`UserDate`) over default content dates (`ContentDate` with `DatePolicy`).
* Supports optional filtering by:
  
  * `block_types` (e.g., `sequential`, `vertical`)
  * `block_keys` (UsageKey or string list)
  * `date_types` (`due`, `start`, etc.)

### Features
* Handles missing schedules and skips invalid or inactive dates gracefully.
* Fully supports both string and object inputs for course and block keys.
* Optimized with prefetching to minimize query overhead.

### Testing
Extensive test suite verifies:

* Correct merging of user and default dates
* All filtering options
* Multiple overrides with the latest one selected
* Handling of edge cases like empty datasets or missing metadata

### CODE REVIEW

**Review by e0d — CHANGES_REQUESTED** [review]

File: edx_when/tests/test_api.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
+"""
+Test cases for the api module of edx-when.
```
[the remark this answers is not in the document]
Comment (kyrylo-kh) [reply edx_when/tests/test_api.py]: Moved test case to main test folder

File: edx_when/api.py [hunk 2]
```diff
@@ -548,6 +570,91 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+                to_attr='user_overrides'
+            )
+        )
+    )
+
+    schedule = get_schedule_for_user(user_id, course_id)
+    end_datetime, cutoff_datetime = _get_end_dates_from_content_dates(content_dates)
+    dates = _resolve_policy_dates(content_dates, schedule, end_datetime, cutoff_datetime)
+
+    for content_date in content_dates:
```
Comment (edschema) [inline edx_when/api.py:632]: Suggestion: Guard against no schedule/ `MissingScheduleError` as in `_resolve_policy_dates` above

File: edx_when/api.py [hunk 3]
```diff
@@ -548,6 +570,91 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+        if content_date.user_overrides:
+            dates[(content_date.location, content_date.field)] = content_date.user_overrides[0].actual_date
```
Comment (edschema) [inline edx_when/api.py]: Question: Is there a guard against shifting the date past a course end date?

File: edx_when/api.py (hunk 2 above)
Comment (kyrylo-kh) [reply edx_when/api.py:632]: Added https://github.com/openedx/edx-when/pull/346/changes/5285469f66b126f38f8b389ae12fc5ad5b3c3c7a

File: edx_when/api.py (hunk 3 above)
Comment (kyrylo-kh) [reply edx_when/api.py]: No guard, by design - matches canonical `get_dates_for_course`, which applies overrides via bare `userdate.actual_date` without end-date capping. Only policy relative dates get capped. Overrides = explicit instructor extensions, allowed past course end

File: edx_when/api.py [hunk 4]
```diff
@@ -548,6 +570,95 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+    Create or update ContentDate entries for a list of assignment objects.
+
+    Arguments:
+        course_key: either a CourseKey or string representation of same
+        assignments: iterable of objects with attributes ``block_key`` (UsageKey) and ``date`` (datetime)
+    """
+    course_key = _ensure_key(CourseKey, course_key)
+    for assignment in assignments:
+        if assignment.date is not None:
+            set_date_for_block(course_key, assignment.block_key, 'due', assignment.date)
```
Comment (farhaanbukhsh) [inline edx_when/api.py]: This should be an atomic operation because if on of the the operation raises errors, others will still go on. What do you think?

File: edx_when/api.py [hunk 5]
```diff
@@ -548,6 +570,95 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+        block_keys: optional list of UsageKey objects or strings to filter by
+        date_types: optional list of date field types to filter by (e.g., ['due', 'start', 'end'])
+
+    Returns:
+        dict with keys as (location, field) tuples and values as date objects
+        User overrides take priority over content defaults
+    """
+    course_id = _ensure_key(CourseKey, course_id)
+
+    content_dates_query = models.ContentDate.objects.filter(
```
Comment (farhaanbukhsh) [inline edx_when/api.py:596]: What about when edx_when.api._are_relative_dates_enabled is true/false, are we expected to return relative deadlines? I was looking into (relative_dates tests)[https://github.com/openedx/edx-when/blob/master/tests/test_api.py#L363], probably we should add this function there as well?

File: edx_when/api.py [hunk 6]
```diff
@@ -548,6 +570,95 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+            Prefetch(
+                'userdate_set',
+                queryset=models.UserDate.objects.filter(user_id=user_id).order_by('-modified'),
+                to_attr='user_overrides'
+            )
+        )
+    )
+
+    schedule = get_schedule_for_user(user_id, course_id)
+    end_datetime, cutoff_datetime = _get_end_dates_from_content_dates(content_dates)
```
Comment (farhaanbukhsh) [inline edx_when/api.py]: Depending on the above answer if I have relative date enable this function can have undesirable behaviour, I have two scenarios in mind.

1. When a relative due date would fall after the course end date, the API should cap   
 it at the course end date. But this function wouldn't do that instead will return the relative due date.

2. For self-paced courses, if the learner started too late, due dates should become    
 unavailable (None) based on the course cutoff logic.

I have tried some tests

```python

    def test_due_filter_still_applies_course_end_cap(self):
        due_block = make_block_id(self.course_key)
        end_block = make_block_id(self.course_key, block_type='course')
        end_date = datetime(2023, 1, 5)
        models.ContentDate.objects.create(
            course_id=self.course_key,
            location=due_block,
            field='due',
            active=True,
            policy=models.DatePolicy.objects.create(rel_date=timedelta(days=10)),
            block_type='sequential'
        )
        models.ContentDate.objects.create(
            course_id=self.course_key,
            location=end_block,
            field='end',
            active=True,
            policy=models.DatePolicy.objects.create(abs_date=end_date),
            block_type='course'
        )
        self._make_enrollment_with_schedule(start_date=datetime(2023, 1, 1), created=datetime(2023, 1, 1))

        result = api.get_user_dates(self.course_key, self.user.id, date_types=['due'])

        assert result == {(due_block, 'due'): end_date}

    def test_due_filter_still_applies_course_cutoff(self):
        due_block = make_block_id(self.course_key)
        end_block = make_block_id(self.course_key, block_type='course')
        end_date = datetime(2023, 1, 20)
        models.ContentDate.objects.create(
            course_id=self.course_key,
            location=due_block,
            field='due',
            active=True,
            policy=models.DatePolicy.objects.create(rel_date=timedelta(days=10)),
            block_type='sequential'
        )
        models.ContentDate.objects.create(
            course_id=self.course_key,
            location=end_block,
            field='end',
            active=True,
            policy=models.DatePolicy.objects.create(abs_date=end_date),
            block_type='course'
        )
        self._make_enrollment_with_schedule(start_date=datetime(2023, 1, 1), created=datetime(2023, 1, 15))

        result = api.get_user_dates(self.course_key, self.user.id, date_types=['due'])
```

**Review by farhaanbukhsh — CHANGES_REQUESTED** [review]
@kyrylo-kh  thank you for the work here, I have added some more comments please have a look.

File: edx_when/api.py [hunk 7]
```diff
@@ -137,6 +137,28 @@ def _get_end_dates_from_content_dates(qset):
     return end_datetime, cutoff_datetime
 
 
+def _resolve_policy_dates(content_dates, schedule=None, end_datetime=None, cutoff_datetime=None):
```
Comment (GlugovGrGlib) [inline edx_when/api.py]: Based on the previous discussion I suggest reusing this helper function inside `def get_dates_for_course` ([lines 270 - 278](https://github.com/openedx/edx-when/pull/346/changes#diff-c3d98af45b99fefbf88d009b7b89c13a1fdea947cb57f3523c1af0c5fa8c4899R270)) as well.
This requires building a policies map inside a helper, and also pass course_id as an additional parameter to normalize cdate using `cdate.location.map_into_course(course_id)` as in `get_dates_for_course` case, but makes more sense in keeping it as a helper method.

Also in case this is done, please update a docstring to describe arguments and new return value.

File: edx_when/api.py (hunk 6 above)
Comment (GlugovGrGlib) [reply edx_when/api.py]: This is a nice catch. Except relative dates case, `_get_end_dates_from_content_dates`  is returning not expected result in case any of the filters are passed, as this method require a course block to correctly calculate end date.
Probably it's necessary to get end date and due directly from DB via ContentDate and then calculate cutoff.

File: edx_when/api.py (hunk 4 above)
Comment (GlugovGrGlib) [reply edx_when/api.py]: It seems to be handled in the follow-up PR already - https://github.com/openedx/edx-when/pull/347/changes#diff-c3d98af45b99fefbf88d009b7b89c13a1fdea947cb57f3523c1af0c5fa8c4899R595.

Maybe in this case, this method can be completely removed from the current PR @kyrylo-kh?

File: edx_when/api.py [hunk 8]
```diff
@@ -548,6 +570,95 @@ def get_schedules_with_due_date(course_id, assignment_date):
...
+            try:
+                actual_date = content_date.user_overrides[0].actual_date
+            except models.MissingScheduleError:
+                continue
+            dates[(content_date.location, content_date.field)] = actual_date
+
+    return dates
+
+
+def get_existing_due_locations(course_key):
```
Comment (GlugovGrGlib) [inline edx_when/api.py]: I suggest improving the function naming, as the current one is a little confusing.
`get_locations_with_due_dates` or `get_blocks_with_due_dates` might suit better

**Review by GlugovGrGlib — CHANGES_REQUESTED** [review]
This looks good, but the current issue around usage of the relative dates feature flag and  `_get_end_dates_from_content_dates` worth addressing before the merge.

File: edx_when/api.py (hunk 6 above)
Comment (farhaanbukhsh) [reply edx_when/api.py]: yup that makes more sense

File: edx_when/api.py [hunk 9]
```diff
@@ -137,6 +137,41 @@ def _get_end_dates_from_content_dates(qset):
...
+    """
+    dates = {}
+    policies = {}
+    for cdate in content_dates:
+        location = cdate.location.map_into_course(course_id) if course_id else cdate.location
+        key = (location, cdate.field)
+        try:
+            dates[key] = cdate.policy.actual_date(schedule, end_datetime, cutoff_datetime)
+        except models.MissingScheduleError:
+            pass
```
Comment (farhaanbukhsh) [inline edx_when/api.py:170]: @kyrylo-kh  these are the lines to be tested, this will increase the coverage and help to test the code better.

**Review by farhaanbukhsh — CHANGES_REQUESTED** [review]
I have added the comment to fix CI, overall the code looks good.

**Review by GlugovGrGlib — APPROVED** [review]
Looks good to me!
Agree about necessity of improving coverage to the target and then this can be merged.

**Comment by farhaanbukhsh** [comment]
@kyrylo-kh Gentle ping on updating this PR 😇

**Comment by kyrylo-kh** [comment]
@farhaanbukhsh hi, thanks for review. We still have a problem with https://github.com/openedx/edx-when/actions/runs/32174982859/job/95834696506?pr=346. Looks like upstream bug with requirements. The same issue in next PR

**Comment by farhaanbukhsh** [comment]
@kyrylo-kh  can you rebase the PR I think we fixed the issue.

**Comment by kyrylo-kh** [comment]
@farhaanbukhsh Done

**Review by e0d — APPROVED** [review]

**Comment by e0d** [comment]
@kyrylo-kh Can you rebase once more.  @GlugovGrGlib thought that some of these commits may been included in the previous PR #347 Is that expected?  It's a little confusing because of the parallel rebasing.  Glib is looking a little more closely and will provide more detail.

**Comment by e0d** [comment]
@farhaanbukhsh can you confirm if your requested change has been handled.

**Comment by farhaanbukhsh** [comment]
@e0d  once @kyrylo-kh  rebases this I will do a final round of testing and will merge it :)

**Review by farhaanbukhsh — APPROVED** [review]
👍

✅ I tested this on master and verified the test cases.
✅ I read through the code
❌  I checked for accessibility issues
❌  Includes documentation
